lib/src/io_utils.dart

```diff
@@ -118,40 +120,96 @@ final RegExp partOfRegexp = RegExp('part of ');
     'as Dartdoc 1.0.0')
 final RegExp newLinePartOfRegexp = RegExp('\npart of ');
 
-/// Best used with Future<void>.
-class MultiFutureTracker<T> {
-  /// Approximate maximum number of simultaneous active Futures.
-  final int parallel;
+typedef TaskQueueClosure<T> = Future<T> Function();
 
-  final Set<Future<T>> _trackedFutures = {};
+class _TaskQueueItem<T> {
+  _TaskQueueItem(this._closure, this._completer, {this.onComplete});
 
-  MultiFutureTracker(this.parallel);
+  final TaskQueueClosure<T> _closure;
+  final Completer<T> _completer;
+  void Function() onComplete;
 
-  /// Wait until fewer or equal to this many Futures are outstanding.
-  Future<void> _waitUntil(int max) async {
-    while (_trackedFutures.length > max) {
-      await Future.any(_trackedFutures);
+  Future<void> run() async {
+    try {
+      _completer.complete(await _closure());
+    } catch (e) {
+      _completer.completeError(e);
+    } finally {
+      onComplete?.call();
     }
   }
+}
+
+/// A task queue of Futures to be completed in parallel, throttling
+/// the number of simultaneous tasks.
+///
+/// The tasks return results of type T.
+class TaskQueue<T> {
+  /// Creates a task queue with a maximum number of simultaneous jobs.
+  /// The [maxJobs] parameter defaults to the number of CPU cores on the
+  /// system.
+  TaskQueue({int maxJobs})
+      : maxJobs = maxJobs ?? io.Platform.numberOfProcessors;
+
+  /// The maximum number of jobs that this queue will run simultaneously.
+  final int maxJobs;
+
+  final Queue<_TaskQueueItem<T>> _pendingTasks = Queue<_TaskQueueItem<T>>();
+  final Set<_TaskQueueItem<T>> _activeTasks = <_TaskQueueItem<T>>{};
+  final Set<Completer<void>> _completeListeners = <Completer<void>>{};
+
+  /// Returns a future that completes when all tasks in the [TaskQueue] are
+  /// complete.
+  Future<void> get tasksComplete {
+    // In case this is called when there are no tasks, we want it to
+    // signal complete immediately.
+    if (_activeTasks.isEmpty && _pendingTasks.isEmpty) {
+      return Future<void>.value();
+    }
+    final completer = Completer<void>();
+    _completeListeners.add(completer);
+    return completer.future;
+  }
+
+  /// Adds a single closure to the task queue, returning a future that
+  /// completes when the task completes.
+  Future<T> add(TaskQueueClosure<T> task) {
+    final completer = Completer<T>();
+    _pendingTasks.add(_TaskQueueItem<T>(task, completer));
+    _processTasks();
```

Comment:
IIUC this seems like it will spawn a lot of _processTasks()  that could potentially do "duplicate" work in the case where the _processTask() from the onComplete has it covered.  This is probably fine though as the duplicates are written to collapse and do nothing.  The old version probably got into trouble doing things cleverly to avoid this...

Reply:
Yes, possibly, but it would also make sure that when a new task is added, it starts running right away.

Reply:
Sure.  I was thinking of the case though when the maximum concurrency is reached and _processTasks() here doesn't have any useful effect.  It seems like it should be OK just to inline _processTasks() here and eliminate an unawaited future that sometimes doesn't do anything.  Your choice, and of course, only if it works just as well.

Reply:
Oh, good point, that makes sense. OK, I inlined it.